Update Nanostack drivers to CMSIS2/RTX5

SeppoTakalo opened the pull request:
* Use statically allocated blocks for mutexes
* Use statically allocated blocks for threads.

**NOTE:** For statically allocated blocks, we are forced to use RTX types and headers, so we are not anymore directly compatible with CMSIS-RTOS2.
But this is how the example documentation does it anyway.

## Status
**READY**

**TODO:**

bulislaw commented on features/FEATURE_COMMON_PAL/nanostack-hal-mbed-cmsis-rtos/ns_event_loop.c:
```diff
@@ -12,12 +14,25 @@
+
+static uint64_t event_thread_stk[MBED_CONF_NANOSTACK_HAL_EVENT_LOOP_THREAD_STACK_SIZE/8];
+static osRtxThread_t event_thread_tcb;
+static const osThreadAttr_t event_thread_attr = {
```
Maybe we could add name for the thread as well, could be useful for debugging.

kjbracey-arm commented:
I'm intensely negative on the fact we apparently are no longer using CMSIS-RTOS2 as an abstraction layer. If it's not an abstraction layer, then what's the point?

This code /was/ portable CMSIS-RTOS code. I'd have written this code using RTX APIs directly in the first place if I'd known that we weren't serious about the abstraction.

But given that rant, yes, this does seem to be doing the job of transforming CMSIS-RTOS code into CMSIS-RTOS2+RTX code correctly.

Some stray printfs?

SeppoTakalo commented:
Just realised that I left too much debugging code there, and did not do the NCS36510 driver.
I'll continue with this.

bulislaw commented:
@kjbracey-arm what do you exactly mean by 'we apparently are no longer using CMSIS-RTOS2 as an abstraction layer'?

kjbracey-arm commented:
We're now apparently having to include RTX header files and use RTX types, as well as the cmsis_os2.h header and types.

The original code would in principle have run on FreeRTOS or anything else providing the CMSIS-RTOS abstraction layer. It's now RTX-only, but just to get the object storage.

This does appear to be a CMSIS-RTOS2 limitation - the docs say that you have to do dynamic allocation to be portable - there's no run-time or compile-time mechanism to find out object size.

"Using the cmsis_os2.h along with dynamic object allocation allows to create source code or libraries that require no modifications when using on a different CMSIS-RTOS2 implementation."

0xc0170 commented:
> I'm intensely negative on the fact we apparently are no longer using CMSIS-RTOS2 as an abstraction layer. If it's not an abstraction layer, then what's the point?

> This code /was/ portable CMSIS-RTOS code. I'd have written this code using RTX APIs directly in the first place if I'd known that we weren't serious about the abstraction.

@kjbracey-arm  good point. I made the exact one today here : https://github.com/ARMmbed/mbed-os/issues/3986#issuecomment-288693726 (why we are including rtx internals to the C++ API). We should review where their abstraction fails - we should discuss this with CMSIS team. Or?

bulislaw commented:
Ah yes, we don't really support RTX dynamic memory or the object pool, mostly not to have to use two carveouts - RTX and HEAP. I think it's better to get the memory size directly from RTX object, rather than use hardcoded constructs like `int32_t _mutex_data[4];`. That being said we talked to Keil guys about ways of solving memory allocation issue, they decided to enable us by declare their allocators as weak so we can overwrite them, possibly wrapping malloc/free. But that should be discussed further and I think it's a task for after the initial update.

kjbracey-arm commented:
Seppo did earlier suggest "just give 64 bytes for each object - bound to be enough".

A slightly more serious suggestion would be to at least abstract the storage as some sort of "cmsis_os2_storage.h", which provides fixed typedefs for each object storage, so you don't end up including RTX headers in every single user of cmsis_os2.h, and having to change all of them if we ever did try to port anything to a different RTOS.  Isolate the RTX include in one place.

(I assume the dynamic allocation idea is also intended to allow you to make CMSIS-RTOS2-portable binary libraries, but we don't care about that - just compile-time portable would be fine).

bulislaw commented:
Yes I agree, it would make everyones life easier. I'll create the header tomorrow and ask you guys for a review.

kjbracey-arm commented:
I do in general like the idea of not requiring the allocation for mutex creation. I do prefer such low-level things to be static, avoiding all sorts of worries about heap allocation or fixed-size pools you need to configure.

But we're kind of working against the grain for CMSIS-RTOS here. The dynamic allocation is a feature for portability, and doing static involves quite a few extra lines of code, plus the magic compile-time knowledge of object size. The extra code isn't a deal-breaker, but that knowledge is.

Conceptually a "cmsis_os2_storage.h" could either just include rtx_lib.h or the single header of whatever OS you knew you were using, as in our case, or it could have knowledge of the maximum of a set of OSes, or it could just go for a "worst case". I think it's something that could be added as a general extension.

SeppoTakalo commented:
Updated the NCS36510 driver also and removed the extra printf()s.

Can I mark this as a done, or should I wait for somebody to come up with the `cmsis_os2_storage.h`

SeppoTakalo commented:
Btw. NCS36510 cannot yet be tested. Segaults at application start.

SeppoTakalo commented:
What work is this requiring?
@bulislaw Are you happy to merge this into your feature branch?

bulislaw commented:
I was hoping to debug this board and test this, before merging this PR. But if you are happy with this as it is without NCS36510, I'm ok with merging.

SeppoTakalo commented:
I don't think the issue is actually related to changes in this PR. Changes are basically identical between all RF drivers that I have done.

* https://github.com/ARMmbed/atmel-rf-driver/tree/feature_cmsis5
* https://github.com/ARMmbed/sal-nanostack-driver-k64f-eth/tree/feature_cmsis5

So there is something else that is causing this board to fail.
We need more testing for this branch anyway.

bulislaw commented:
Ok, merging. I'll try to get this board (we don't have it in the core team but maybe apps will have one i can borrow) and try to debug it.
